File: Depth-First Search/Combinations/String/LET_no301_Remove Invalid Parentheses.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        lefts, rights = self.to_remove(s)
        res = []
        self.helper(s, 0, lefts, rights, [], res)
        return res

    def helper(self, s, start, lefts, rights, temp, res):
        if lefts == 0 and rights == 0:
            cur = ''.join(s[i:j] for i, j in temp) + s[start:]
            if self.is_valid(cur):
                res.append(cur)
            return

        if lefts < 0 or rights < 0:
            return

        for end in range(start, len(s)):
            if end > start and s[end] == s[end - 1]:
                continue

            if s[end] == '(':
                temp.append((start, end))
                self.helper(s, end + 1, lefts - 1, rights, temp, res)
                temp.pop()
            elif s[end] == ')':
                temp.append((start, end))
                self.helper(s, end + 1, lefts, rights - 1, temp, res)
                temp.pop()

    def to_remove(self, s):
        lefts, rights = 0, 0
        for c in s:
            if c == '(':
                lefts += 1
            elif c == ')':
                if lefts > 0:
                    lefts -= 1
                else:
                    rights += 1
        
        return lefts, rights
    
    def is_valid(self, s):
        lefts, rights = self.to_remove(s)
        return lefts == 0 and rights == 0

File: Depth-First Search/Combinations/String/LET_no301_Remove Invalid Parentheses.cpp
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        auto p = toRemove(s);
        int lefts = p.first, rights = p.second;
        vector<string> temp, res;
        helper(s, 0, lefts, rights, temp, res);
        return res;
    }

    pair<int, int> toRemove(string& s) {
        int lefts = 0, rights = 0;
        for (char c: s) {
            if (c == '(')
                lefts++;
            else if (c == ')') {
                if (lefts > 0)
                    lefts--;
                else
                    rights++;
            }
        }

        return {lefts, rights};
    }

    void helper(string& s, int start, int lefts, int rights, vector<string>& temp, vector<string>& res) {
        if (lefts == 0 && rights == 0) {
            string cur = "";
            for (string& sub : temp)
                cur += sub;
            cur += s.substr(start);
            if (isValid(cur))
                res.push_back(cur);
            return;
        }

        if (lefts < 0 || rights < 0)
            return;

        for (int end = start; end < s.size(); end++) {
            if (end > start && s[end] == s[end - 1])
                continue;

            if (s[end] == '(') {
                temp.push_back(s.substr(start, end - start));
                helper(s, end + 1, lefts - 1, rights, temp, res);
                temp.pop_back();
            }
            else if (s[end] == ')') {
                temp.push_back(s.substr(start, end - start));
                helper(s, end + 1, lefts, rights - 1, temp, res);
                temp.pop_back();
            }
        }
    }

    bool isValid(string& s) {
        auto p = toRemove(s);
        return p.first == 0 && p.second == 0;
    }
};